src/esp/scene/SceneNode.cpp

```diff
@@ -16,16 +18,61 @@ SceneNode::SceneNode(SceneNode& parent)
   setId(parent.getId());
   setCachedTransformations(Mn::SceneGraph::CachedTransformation::Absolute);
   absoluteTransformation_ = absoluteTransformation();
+  nodeSensorSuite_ = new esp::sensor::SensorSuite(*this);
+  subtreeSensorSuite_ = new esp::sensor::SensorSuite(*this);
 }
 
 SceneNode::SceneNode(MagnumScene& parentNode)
     : Mn::SceneGraph::AbstractFeature3D{*this} {
   setParent(&parentNode);
   setCachedTransformations(Mn::SceneGraph::CachedTransformation::Absolute);
   absoluteTransformation_ = absoluteTransformation();
+  nodeSensorSuite_ = new esp::sensor::SensorSuite(*this);
+  subtreeSensorSuite_ = new esp::sensor::SensorSuite(*this);
+}
+
+SceneNode::~SceneNode() {
+  // if the entire scene graph is being deleted, no need to update anything
+  if (SceneGraph::isRootNode(*this)) {
+    return;
+  }
+  // if it is to delete a subtree
+  // then two cases:
+  // 1) current node is the root node (of this subtree).
+  // In case 1, you do not have to do anything since its parent node is nullptr
+  // already!! This is because magnum destroys the nodes in a top-down manner
+  // recursively, root node first, and then its child nodes
+  if (!this->parent()) {
+    return;
+  }
+
+  // 2) current node is NOT the root node.
+  // In case 2, you need update the sensorSuites stored in each ancestor node in
+  // a bottom-up manner.
+
+  // First erase all nodes in nodeSensorSuite from its parent's nodeSensorSuite
+  for (const auto& sensor : nodeSensorSuite_->getSensors()) {
+    static_cast<SceneNode*>(this->parent())
+        ->getNodeSensorSuite()
+        .getSensors()
+        .erase(sensor.first);
+  }
+
+  // go bottom to the top, and erase sensors and all subtreeSensors from its
+  // ancestors' subtreeSensorSuites
+  for (const auto& sensor : subtreeSensorSuite_->getSensors()) {
+    SceneNode* currentNode = this;
+    do {
+      currentNode = dynamic_cast<SceneNode*>(currentNode->parent());
+      // no need to worry that currentNode could be nullptr, no chance
+      currentNode->getSubtreeSensorSuite().getSensors().erase(sensor.first);
+    } while (!SceneGraph::isRootNode(*currentNode));
+  }
 }
 
 SceneNode& SceneNode::createChild() {
```

Comment:
Also upgrade this function to:
`createChild(SceneNodeTag childNodeTag = {})`
So when you create sensors in the SensorFactory, you can directly set the tag even earlier during the construction.